index.js

```diff
@@ -132,11 +134,19 @@ Router.prototype._insert = function (method, path, kind, params, handler, store)
 }
 
 Router.prototype.lookup = function (req, res) {
+  // strip the url from querystring and hashes
   var i = 0
   var len = req.url.length
   while (i < len && req.url[i] !== '?' && req.url[i] !== '#') i++
-  var handle = this.find(req.method, req.url.slice(0, i))
+
+  // search the route inside the cache
+  var handle = this.cache.get(req.url.slice(0, i))
+  if (handle) return handle.handler(req, res, handle.params, handle.store)
+
+  // if the route in not saved in the cache, we search it
+  handle = this.find(req.method, req.url.slice(0, i))
   if (!handle) return this._defaultRoute(req, res)
+  this.cache.set(req.url.slice(0, i), handle)
```

Comment:
You are right, if I run this benchmark
```js
.add('lookup dynamic route - cache random path', function () {
  findMyWay.lookup({ method: 'GET', url: '/user/' + i++ }, null)
})
```
I got a 8x-10x perf loss.

We can make the cache optional and warn the user about this behaviour. So if the user will have a lot of static routes can use the caching system, otherwise he can disable it.